service/entityresolution/keycloak/v2/retrieve.go

```diff
@@ -0,0 +1,120 @@
+	var cacheKey string
+	switch {
+	case getUserParams.Email != nil:
+		cacheKey = fmt.Sprintf("%s::user::%s", realm, *getUserParams.Email)
+	case getUserParams.Username != nil:
+		cacheKey = fmt.Sprintf("%s::user::%s", realm, *getUserParams.Username)
+	default:
+		return nil, errors.New("either email or username must be provided")
+	}
+
+	retrievalFunc := func() ([]*gocloak.User, error) {
+		return connector.client.GetUsers(ctx, connector.token.AccessToken, realm, getUserParams)
+	}
+	users, err := retrieveWithKey[[]*gocloak.User](ctx, cacheKey, svcCache, logger, retrievalFunc)
+	if err != nil {
+		return nil, err
+	}
+	return users, nil
+}
+
+func retrieveGroupsByEmail(ctx context.Context, logger *logger.Logger, groupEmail string, realm string, svcCache *cache.Cache, connector *Connector) ([]*gocloak.Group, error) {
+	cacheKey := fmt.Sprintf("%s::group::%s", realm, groupEmail)
+	retrievalFunc := func() ([]*gocloak.Group, error) {
+		return connector.client.GetGroups(
+			ctx,
+			connector.token.AccessToken,
+			realm,
+			gocloak.GetGroupsParams{Search: func() *string { t := groupEmail; return &t }()},
+		)
+	}
+	group, err := retrieveWithKey[[]*gocloak.Group](ctx, cacheKey, svcCache, logger, retrievalFunc)
+	if err != nil {
+		return nil, err
+	}
+	return group, nil
+}
+
+func retrieveGroupByID(ctx context.Context, logger *logger.Logger, groupID string, realm string, svcCache *cache.Cache, connector *Connector) (*gocloak.Group, error) {
+	cacheKey := fmt.Sprintf("%s::group::%s", realm, groupID)
+	retrievalFunc := func() (*gocloak.Group, error) {
+		return connector.client.GetGroup(ctx, connector.token.AccessToken, realm, groupID)
+	}
+	group, err := retrieveWithKey[*gocloak.Group](ctx, cacheKey, svcCache, logger, retrievalFunc)
+	if err != nil {
+		return nil, err
+	}
+	return group, nil
+}
+
+func retrieveGroupMembers(ctx context.Context, logger *logger.Logger, groupID string, realm string, svcCache *cache.Cache, connector *Connector) ([]*gocloak.User, error) {
+	cacheKey := fmt.Sprintf("%s::group::%s::members", realm, groupID)
+	retrievalFunc := func() ([]*gocloak.User, error) {
+		return connector.client.GetGroupMembers(ctx, connector.token.AccessToken, realm, groupID, gocloak.GetGroupsParams{})
+	}
+	members, err := retrieveWithKey[[]*gocloak.User](ctx, cacheKey, svcCache, logger, retrievalFunc)
+	if err != nil {
+		return nil, err
+	}
+	return members, nil
+}
+
+func retrieveWithKey[T any](ctx context.Context, cacheKey string, svcCache *cache.Cache, logger *logger.Logger, retrieveFunc func() (T, error)) (T, error) {
+	if svcCache != nil {
+		cachedData, err := svcCache.Get(ctx, cacheKey)
+		switch {
+		case err == nil:
+			if retrieved, ok := cachedData.(T); ok {
+				return retrieved, nil
+			}
+			logger.Error("cache data type assertion failed")
+		case errors.Is(err, cache.ErrCacheMiss):
+			logger.Debug("cache miss", slog.String("key", cacheKey))
```

Comment:
Is this log a duplicate because we log in pkg/cache when we have a cache miss? https://github.com/opentdf/platform/blob/c16361c8048a6685fa7781b6f1f403f13c897792/service/pkg/cache/cache.go#L99